core/Updates/5.0.0-b1.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link    https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+
+namespace Piwik\Updates;
+
+use Piwik\Db;
+use Piwik\Common;
+use Piwik\Updater;
+use Piwik\Updater\Migration\Db as DbAlias;
+use Piwik\Updater\Migration\Db\DropIndex;
+use Piwik\Updater\Migration\Db\Sql;
+use Piwik\Updates as PiwikUpdates;
+
+/**
+ * Update for version 5.0.0-b1
+ */
+class Updates_5_0_0_b1 extends PiwikUpdates
+{
+    private $tableName;
+    private $indexName;
+
+    public function __construct()
+    {
+        $this->tableName = Common::prefixTable('log_visit');
+        $this->indexName = 'index_idsite_idvisitor';
+    }
+
+    public function doUpdate(Updater $updater)
+    {
+        if ($this->requiresUpdatedLogVisitTableIndex()) {
+            $this->updateLogVisitTableIndex();
+        }
+    }
+
+    private function updateLogVisitTableIndex()
+    {
+        try {
+            $dropIndex = new DropIndex($this->tableName, $this->indexName);
+            $dropIndex->exec();
+
+            // Using the base `Sql` class instead of the AddIndex class as it doesn't support DESC collation
+            $sql = "ALTER TABLE `{$this->tableName}` ADD INDEX `{$this->indexName}` (`idsite`, `idvisitor`, `visit_last_action_time` DESC)";
+            $addIndex = new Sql($sql, [DbAlias::ERROR_CODE_DUPLICATE_KEY, DbAlias::ERROR_CODE_KEY_COLUMN_NOT_EXISTS]);
+            $addIndex->exec();
+        } catch (\Exception $e) {
+            if (!$dropIndex->shouldIgnoreError($e)) {
+                throw $e;
+            }
+        }
```

Comment:
Is there a reason why this isn't done using migrations like e.g. here:
https://github.com/matomo-org/matomo/blob/0e7530e93f0ada22024e7226badc1d94c8309f3c/core/Updates/4.0.0-rc3.php#L20-L44

Using migrations makes it possible for the user, to see the queries before running the update (and run them manually instead)

Reply:
@sgiehl Yes, the AddIndex class wouldn't allow me to easily use the `DESC` collation with the `visit_last_action_time` column. I'm open to ideas if you know of any methods to make this work?

Reply:
You can still use a custom sql with migrations. e.g. `$this->migration->db->sql('....')`